### PR TITLE
[compiler-rt] Allow running tests without installing first

### PR DESCRIPTION
**Pull request by arichardson**
Currently, the testsuite uses the default runtimes path to find the
runtimes libraries which may or may not match the just-built runtimes.
This change uses the `-resource-dir` flag for clang whenever
`COMPILER_RT_TEST_STANDALONE_BUILD_LIBS` is set to ensure that we are
actually testing the currently built libraries rather than the ones
bundled with `${COMPILER_RT_TEST_COMPILER}`.

The existing logic works fine when clang and compiler-rt share the same
build directory ``-DLLVM_ENABLE_PROJECTS=clang;compiler-rt`, but when
building compiler-rt separately we need to tell the compiler used for
the tests where it can find the just-built libraries.

This reduces the fixes check-all failures to one in my configuration:
```
cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo -G Ninja
-DCMAKE_C_COMPILER=$HOME/output/upstream-llvm/bin/clang
-DCMAKE_CXX_COMPILER=$HOME/output/upstream-llvm/bin/clang++
-DCOMPILER_RT_INCLUDE_TESTS=ON
-DLLVM_EXTERNAL_LIT=$HOME/build/upstream-llvm-project-build/bin/llvm-lit
-DLLVM_CMAKE_DIR=$HOME/output/upstream-llvm
-DCOMPILER_RT_DEBUG=OFF
-S $HOME/src/upstream-llvm-project/compiler-rt
-B $HOME/src/upstream-llvm-project/compiler-rt/cmake-build-all-sanitizers
```

### CODE REVIEW

File: compiler-rt/CMakeLists.txt
Comment (delcypher) [inline compiler-rt/CMakeLists.txt:594]: Is this left over from debugging? This warning message doesn't seem that helpful to a developer.

File: compiler-rt/test/lit.common.cfg.py
Comment (delcypher) [inline compiler-rt/test/lit.common.cfg.py:180]: Sorry. It's been a while since I looked at this code. If you use a different resource directory why can't the compiler use the header files in that resource directory?

File: compiler-rt/test/lit.common.cfg.py
Comment (delcypher) [inline compiler-rt/test/lit.common.cfg.py:188]: The flags below create kind of a odd mix

* Some external clang that has its own independent resource directory.
* The external clang is being forced to use a different resource directory than what it was shipped with
* The external clang is being forced to use a different set of headers than what it was shipped with

I'm not sure that combination is guaranteed to work given that compiler-rt can be tightly coupled with clang.

If I've read this code correctly the change you've made is on the common path (because `COMPILER_RT_TEST_STANDALONE_BUILD_LIBS` is `ON` by default) but AFAIK your change was never needed before because normally you would build compiler-rt with a just built clang which I think should mean there's no difference between what `-print-resource-dir` reports and `config.compiler_rt_obj_root`. Or have I misunderstood something?

**Review by delcypher — COMMENTED** [review]
What is the `[𝘀𝗽𝗿] initial version` commit message about?

File: compiler-rt/test/lit.common.cfg.py
Comment (delcypher) [reply compiler-rt/test/lit.common.cfg.py:188]: @arichardson To be clear. I'm not necessarily against your change but I'm a bit concerned about your change being on the default path given that it's not normally necessary.

File: compiler-rt/CMakeLists.txt
Comment (arichardson) [reply compiler-rt/CMakeLists.txt:594]: Yes, will remove!

File: compiler-rt/test/lit.common.cfg.py
Comment (arichardson) [reply compiler-rt/test/lit.common.cfg.py:180]: I'm overriding the resource directory to point at the compiler-rt build directory. Unlike the clang install dir it won't contain the builtin headers (e.g. stdint.h, stddef.h, etc.). I could copy those files, but using this approach is less error prone.

File: compiler-rt/test/lit.common.cfg.py
Comment (arichardson) [reply compiler-rt/test/lit.common.cfg.py:188]: > The flags below create kind of a odd mix
> 
> * Some external clang that has its own independent resource directory.
In my case it is actually a clang built from the same monorepo commit hash, but it should also work with a system-provided clang.
> * The external clang is being forced to use a different resource directory than what it was shipped with
The intention of this change is to test the libraries that we actually built just now rather than whatever happened to be installed to clang's resource directory. There is absolutely no guarantee that those match.
> * The external clang is being forced to use a different set of headers than what it was shipped with
The whole point of this extra `-idirafter` is to make sure that the external clang uses its set of headers (the resource_dir variable points to the non-overriden directory).

> 
> I'm not sure that combination is guaranteed to work given that compiler-rt can be tightly coupled with clang.
> 
> If I've read this code correctly the change you've made is on the common path (because `COMPILER_RT_TEST_STANDALONE_BUILD_LIBS` is `ON` by default) but AFAIK your change was never needed before because normally you would build compiler-rt with a just built clang which I think should mean there's no difference between what `-print-resource-dir` reports and `config.compiler_rt_obj_root`. Or have I misunderstood something?

The resource directories will only match if you build everything in on go (I believe with `LLVM_ENABLE_PROJECTS=clang;compiler-rt`). However, I am installing clang first and then building *only* compiler-rt (using the cmake invocation listed above), so the clang resource directory points to where clang was installed.

I order to not affect the all-in-one build I will update the check to not add flags if the resource-dir matches the compiler_rt_objdir. Hopefully that addresses your concerns.

**Comment by arichardson** [comment]
> What is the `[𝘀𝗽𝗿] initial version` commit message about?

I'm trying to use getcord/spr for stacked pull requests and it appears this is what it does to the commit messages when uploading them.

File: compiler-rt/test/lit.common.cfg.py
Comment (arichardson) [reply compiler-rt/test/lit.common.cfg.py]: Logic has been updated now to only add these flags when using the non-LLVM_ENABLE_PROJECTS=clang;compiler-rt build.

File: compiler-rt/test/lit.common.cfg.py
Comment (arichardson) [reply compiler-rt/test/lit.common.cfg.py]: Clarified comment, should hopefully address this.

**Comment by MaskRay** [comment]
> wants to merge 2 commits into [users/arichardson/spr/main.compiler-rt-allow-running-tests-without-installing-first](https://github.com/llvm/llvm-project/tree/users/arichardson/spr/main.compiler-rt-allow-running-tests-without-installing-first)

Note: since the base branch is not `main`, "Squash and merge" will not merge the patch to "main", but you can manually push it to `main` after approved. You can still click "Squash and merge" so that the PR shows "merged".

**Review by MaskRay — APPROVED** [review]
I have a regular clang build at /tmp/Rel. If I create a standalone compiler-rt build at /tmp/out/crt, I see that with this PR, clang invoked by llvm-lit will get `-resource-dir=/tmp/out/crt -Wl,-rpath,/tmp/out/crt/lib/linux`, which utilizes `include/sanitizer/*_interface.h` (used by certain asan tests) and runtime libraries.

```
#include <sanitizer/allocator_interface.h>
#include "sanitizer/asan_interface.h"
```

**Comment by arichardson** [comment]
> > wants to merge 2 commits into [users/arichardson/spr/main.compiler-rt-allow-running-tests-without-installing-first](https://github.com/llvm/llvm-project/tree/users/arichardson/spr/main.compiler-rt-allow-running-tests-without-installing-first)
> 
> Note: since the base branch is not `main`, "Squash and merge" will not merge the patch to "main", but you can manually push it to `main` after approved. You can still click "Squash and merge" so that the PR shows "merged".

Thanks, trying to get my head around using spr, but I'm a bit scared of `spr land`, so I will probably edit the base branch of the PR in the web UI and land from there.

File: compiler-rt/test/lit.common.cfg.py
Comment (MaskRay) [inline compiler-rt/test/lit.common.cfg.py:198]: Can `target_cflags` be made to a `list` instead of space-separate `str`?

Comment (arichardson) [reply compiler-rt/test/lit.common.cfg.py:198]: I agree that this would be cleaner, I will submit an independent cleanup PR.

**Review by vitalybuka — APPROVED** [review]
Looks like there are a bunch of CMAKE related patches from you.
If possible, please try to land them with delay in-between for easier reverts if needed.

File: compiler-rt/test/lit.common.cfg.py
Comment (delcypher) [inline compiler-rt/test/lit.common.cfg.py:188]: Nit: If `config.test_standalone_build_libs` and `config.compiler_id` `!=` `Clang` we may want to emit a warning/error about an unsupported compiler.

**Review by delcypher — APPROVED** [review]
Thanks for addressing my feedback. I have a minor nit that you can address if you feel its worth it. When landing please make sure you squash your commits into a single commit and rewrite the commit message to be something that conforms to LLVM's style for commit messages.

File: compiler-rt/test/lit.common.cfg.py
Comment (arichardson) [reply compiler-rt/test/lit.common.cfg.py:188]: That absolutely makes sense and I'll add that warning. Using GCC currently already fails with obscure errors later, but having a warning before should make that less surprising :)

**Comment by arichardson** [comment]
> Looks like there are a bunch of CMAKE related patches from you. If possible, please try to land them with delay in-between for easier reverts if needed.

Thanks, I'll make sure to leave enough delays for buildbots to shout at me when landing these changes :)

**Comment by jplehr** [comment]
Hi. This may have caused a breakage on one of our buildbots https://lab.llvm.org/buildbot/#/builders/165/builds/51837
Let me know if you need some assistance.

**Comment by dyung** [comment]
We are also seeing similar problems with our internal builder after this commit:
```
FAILED: projects/compiler-rt/lib/memprof/tests/MemProfUnitTests 
: && /usr/lib/ccache/g++ -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -fno-lifetime-dse -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-uninitialized -Wno-nonnull -Wno-class-memaccess -Wno-redundant-move -Wno-pessimizing-move -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wno-misleading-indentation -fdiagnostics-color -ffunction-sections -fdata-sections -Wall -Wno-unused-parameter -O3 -DNDEBUG -fuse-ld=gold    -resource-dir=/home/jenkins/j/w/workspace/opensource/opensource_build/build/./lib/../lib/clang/19 -Wl,-rpath,/home/jenkins/j/w/workspace/opensource/opensource_build/build/./lib/../lib/clang/19/lib/x86_64-unknown-linux-gnu -pthread <snip>
g++: error: unrecognized command line option ‘-resource-dir=/home/jenkins/j/w/workspace/opensource/opensource_build/build/./lib/../lib/clang/19’
```

**Comment by saiislam** [comment]
Hey @arichardson 

I had to just revert this patch with [2084a07](https://github.com/llvm/llvm-project/commit/2084a07087a55b55bb3c2a8aafbe1c4464fdf796) because it was breaking the buildbot as mentioned by @jplehr earlier along with our local builds.

In both cases, the error was the same : `unrecognized command-line option ‘-resource-dir=`

Let me or @jplehr know if we can assist you in anyway.

File: compiler-rt/test/lit.common.cfg.py
Comment (mstorsjo) [inline compiler-rt/test/lit.common.cfg.py:205]: In addition to the fix for Windows that was made in 10b1864dff816174cd83fb2d3bc622e25fcf0f8a, this bit here also would need a similar condition - otherwise testing fails with the same error, see e.g. https://github.com/mstorsjo/llvm-mingw/actions/runs/8592846078/job/23549755414.

I see that this whole PR was reverted later, but when preparing for a reland, make sure to conditionalize this case of `-Wl,-rpath` as well.

**Comment by arichardson** [comment]
> We are also seeing similar problems with our internal builder after this commit:
> 
> ```
> FAILED: projects/compiler-rt/lib/memprof/tests/MemProfUnitTests 
> : && /usr/lib/ccache/g++ -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -fno-lifetime-dse -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-uninitialized -Wno-nonnull -Wno-class-memaccess -Wno-redundant-move -Wno-pessimizing-move -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wno-misleading-indentation -fdiagnostics-color -ffunction-sections -fdata-sections -Wall -Wno-unused-parameter -O3 -DNDEBUG -fuse-ld=gold    -resource-dir=/home/jenkins/j/w/workspace/opensource/opensource_build/build/./lib/../lib/clang/19 -Wl,-rpath,/home/jenkins/j/w/workspace/opensource/opensource_build/build/./lib/../lib/clang/19/lib/x86_64-unknown-linux-gnu -pthread <snip>
> g++: error: unrecognized command line option ‘-resource-dir=/home/jenkins/j/w/workspace/opensource/opensource_build/build/./lib/../lib/clang/19’
> ```

That sounds like a either a misconfigured build-bot or a bug in compiler-rt tests using the host compiler instead of the test compiler for this target? This flag will only be added if COMPILER_RT_TEST_COMPILER_ID==Clang, so it definitely should not be added for GCC.

File: compiler-rt/test/lit.common.cfg.py
Comment (arichardson) [reply compiler-rt/test/lit.common.cfg.py:205]: That makes sense. Unfortunately I don't have a way of testing this on Windows, and I was hoping the pre-commit CI would have caught it but I guess that does not include mingw.

**Comment by dyung** [comment]
> > We are also seeing similar problems with our internal builder after this commit:
> > ```
> > FAILED: projects/compiler-rt/lib/memprof/tests/MemProfUnitTests 
> > : && /usr/lib/ccache/g++ -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -fno-lifetime-dse -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-uninitialized -Wno-nonnull -Wno-class-memaccess -Wno-redundant-move -Wno-pessimizing-move -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wno-misleading-indentation -fdiagnostics-color -ffunction-sections -fdata-sections -Wall -Wno-unused-parameter -O3 -DNDEBUG -fuse-ld=gold    -resource-dir=/home/jenkins/j/w/workspace/opensource/opensource_build/build/./lib/../lib/clang/19 -Wl,-rpath,/home/jenkins/j/w/workspace/opensource/opensource_build/build/./lib/../lib/clang/19/lib/x86_64-unknown-linux-gnu -pthread <snip>
> > g++: error: unrecognized command line option ‘-resource-dir=/home/jenkins/j/w/workspace/opensource/opensource_build/build/./lib/../lib/clang/19’
> > ```
> 
> That sounds like a either a misconfigured build-bot or a bug in compiler-rt tests using the host compiler instead of the test compiler for this target? This flag will only be added if COMPILER_RT_TEST_COMPILER_ID==Clang, so it definitely should not be added for GCC.

It was working before your change, and it is a similar failure as what @jplehr's buildbot reported (they seem to be using `/usr/bin/c++` as the compiler, not clang.

**Comment by vitalybuka** [comment]
Build bot still uses '-DLLVM_ENABLE_PROJECTS=clang;lld;clang-tools-extra;compiler-rt' 

can be updated here https://github.com/llvm/llvm-zorg/blob/3ed3d915a9d4de7967dd20a35a9a266b09cdba86/zorg/buildbot/builders/annotated/hip-build.sh#L4

**Comment by arichardson** [comment]
> > > We are also seeing similar problems with our internal builder after this commit:
> > > ```
> > > FAILED: projects/compiler-rt/lib/memprof/tests/MemProfUnitTests 
> > > : && /usr/lib/ccache/g++ -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -fno-lifetime-dse -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-uninitialized -Wno-nonnull -Wno-class-memaccess -Wno-redundant-move -Wno-pessimizing-move -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wno-misleading-indentation -fdiagnostics-color -ffunction-sections -fdata-sections -Wall -Wno-unused-parameter -O3 -DNDEBUG -fuse-ld=gold    -resource-dir=/home/jenkins/j/w/workspace/opensource/opensource_build/build/./lib/../lib/clang/19 -Wl,-rpath,/home/jenkins/j/w/workspace/opensource/opensource_build/build/./lib/../lib/clang/19/lib/x86_64-unknown-linux-gnu -pthread <snip>
> > > g++: error: unrecognized command line option ‘-resource-dir=/home/jenkins/j/w/workspace/opensource/opensource_build/build/./lib/../lib/clang/19’
> > > ```
> > 
> > 
> > That sounds like a either a misconfigured build-bot or a bug in compiler-rt tests using the host compiler instead of the test compiler for this target? This flag will only be added if COMPILER_RT_TEST_COMPILER_ID==Clang, so it definitely should not be added for GCC.
> 
> It was working before your change, and it is a similar failure as what @jplehr's buildbot reported (they seem to be using `/usr/bin/c++` as the compiler, not clang).

Should be fixed by https://github.com/llvm/llvm-project/pull/88074. This was a pre-existing bug that just happened to be exposed by this change.

**Comment by dyung** [comment]
> > > > We are also seeing similar problems with our internal builder after this commit:
> > > > ```
> > > > FAILED: projects/compiler-rt/lib/memprof/tests/MemProfUnitTests 
> > > > : && /usr/lib/ccache/g++ -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -fno-lifetime-dse -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-uninitialized -Wno-nonnull -Wno-class-memaccess -Wno-redundant-move -Wno-pessimizing-move -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wno-misleading-indentation -fdiagnostics-color -ffunction-sections -fdata-sections -Wall -Wno-unused-parameter -O3 -DNDEBUG -fuse-ld=gold    -resource-dir=/home/jenkins/j/w/workspace/opensource/opensource_build/build/./lib/../lib/clang/19 -Wl,-rpath,/home/jenkins/j/w/workspace/opensource/opensource_build/build/./lib/../lib/clang/19/lib/x86_64-unknown-linux-gnu -pthread <snip>
> > > > g++: error: unrecognized command line option ‘-resource-dir=/home/jenkins/j/w/workspace/opensource/opensource_build/build/./lib/../lib/clang/19’
> > > > ```
> > > 
> > > 
> > > That sounds like a either a misconfigured build-bot or a bug in compiler-rt tests using the host compiler instead of the test compiler for this target? This flag will only be added if COMPILER_RT_TEST_COMPILER_ID==Clang, so it definitely should not be added for GCC.
> > 
> > 
> > It was working before your change, and it is a similar failure as what @jplehr's buildbot reported (they seem to be using `/usr/bin/c++` as the compiler, not clang).
> 
> Should be fixed by #88074. This was a pre-existing bug that just happened to be exposed by this change.

Thanks! I'm also updating our internal builder to use `-DLLVM_ENABLE_RUNTIMES` as @vitalybuka indicated. Our bot was setup a long time ago and seems to have gotten out of date.

**Pull request merged**
